src/porepy/compositional/base.py

```diff
@@ -0,0 +1,855 @@
+        """The specific volume of the fluid mixture as the reciprocal of
+        :attr:`density`."""
+
+        self.specific_enthalpy: Callable[[pp.SubdomainsOrBoundaries], pp.ad.Operator]
+        """Specific enthalpy of the fluid mixture.
+
+        Its thermodynamically consistent representation is
+        :math:`\\sum_j y_j h_j`, with :math:`y_j` being :attr:`Phase.fraction` and
+        :math:`h_j` being :attr:`Phase.specific_enthalpy`
+
+        """
+
+    def __str__(self) -> str:
+        """
+        Returns:
+            A string representation of the composition, with information about present
+            components and phases.
+
+        """
+        out = f"Fluid mixture with {self.num_components} components:"
+        for component in self.components:
+            out += f"\n\t{component.name}"
+        out += f"\nand {self.num_phases} phases:"
+        for phase in self.phases:
+            out += f"\n\t{phase.name}"
+        return out
+
+    @property
+    def num_components(self) -> int:
+        """Number of components in this mixture."""
+        return len(self._components)
+
+    @property
+    def num_phases(self) -> int:
+        """Number of phases phases in this mixture."""
+        return len(self._phases)
+
+    @property
+    def num_equilibrium_equations(self) -> int:
+        """Number of necessary equilibrium equations for this composition, based on the
+        number of added components and phases ``num_components * (num_phases - 1)``.
+
+        Equilibrium equation (isofugacity constraints) are always formulated w.r.t. the
+        reference phase, hence ``num_phases - 1``.
+
+        """
+        return self.num_components * (self.num_phases - 1)
+
+    @property
+    def components(self) -> Generator[Component, None, None]:
+        """
+        Note:
+            The first component is always the reference component.
+
+        Yields:
+            Components added to the composition.
+
+        """
+        for C in self._components:
+            yield C
+
+    @property
+    def phases(self) -> Generator[Phase, None, None]:
+        """
+        Note:
+            The first phase is always the reference phase.
+
+        Yields:
+            Phases modelled by the composition class.
+
+        """
+        for P in self._phases:
+            yield P
+
+    @property
+    def reference_phase(self) -> Phase:
+        """Returns the reference phase.
+
+        As of now, the first **non-gas-like** phase is declared as reference
+        phase (based on :attr:`~porepy.compositional.phase.Phase.gaslike`).
+
+        The reference phase is always the first element in :meth:`phases`.
+
+        The implications of a phase being the reference phase include:
+
+        - The reference phase :attr:`~Phase.fraction` and :attr:`~Phase.saturation` can
+          be eliminated by unity and are **not** independent variables.
+
+        """
+        return self._phases[0]
```

Comment:
Makes sense, yes.
But maybe we could do it by re-ordering the phases when the user sets the index? Otherwise many methods of the compiled code which will come later will need an additional argument (this index).

Right now, the user can choose which liquid-like phase can is determined as reference phase, by changing their order when creating a mixture (argument ``phases``)

Reply:
Added a settable property to change index of reference phase and components